C error starting V on macOS older than 10.12 due to clock_gettime missing

Hi, I don't know if you want to support macOS earlier than 10.12.  If you don't, perhaps the easiest way to make this issue go away is to declare that the macOS version of V requires 10.12 or higher.

Versions of macOS before 10.12 don't have `clock_gettime`, you had to use a workaround like https://gist.github.com/jbenet/1087739 so when I downloaded V for macOS version 0.1.28 from vlang.io and tried it on an older Mac Mini running Mac OS X 10.7.5 I got this:

    mac:/tmp$ unzip v_macos.zip
    ...
    mac:/tmp$ cd v
    mac:/tmp/v$ ./v
    For usage information, quit V REPL and run `v help`
    cannot compile `/private/tmp/v/cmd/tools/vrepl.v`:
    ==================
                                                                        ^
    /var/folders/ng/l1tjl5bx13ncfqq6h_6h7qy40000gn/T/v/vrepl.tmp.c:11319:2: error: implicit declaration of function 'clock_gettime' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        clock_gettime(CLOCK_MONOTONIC, &ts);
        ^
    /var/folders/ng/l1tjl5bx13ncfqq6h_6h7qy40000gn/T/v/vrepl.tmp.c:11319:16: error: use of undeclared identifier 'CLOCK_MONOTONIC'
        clock_gettime(CLOCK_MONOTONIC, &ts);
                      ^
    /var/folders/ng/l1tjl5bx13ncfqq6h_6h7qy40000gn/T/v/vrepl.tmp.c:11333:16: error: use of undeclared identifier 'CLOCK_REALTIME'
        clock_gettime(CLOCK_REALTIME, &ts);
                      ^
    /var/folders/ng/l1tjl5bx13ncfqq6h_6h7qy40000gn/T/v/vrepl.tmp.c:12952:83: warning: missing field 'c_oflag' initializer [-Wmissing-field-initializers]
        readline__Readline readline = (readline__Readline){.is_raw = 0,.orig_termios = {0},.current = {0},.cursor = 0,.overwrite = 0,.cursor_row_offset = 0,.prompt = (string){.str=(byteptr)""},.prompt_offset = 0,.previous_lines = __new_array(0, 1, sizeof(ustring)),.search_index = 0,.is_tty = 0,};
    ...
    ==================
    (Use `v -cg` to print the entire error message)
    
    builder error:
    ==================
    C error. This should never happen.
    
    If you were not working with C interop, please raise an issue on GitHub:
    
    https://github.com/vlang/v/issues/new/choose
    
    You can also use #help on Discord: https://discord.gg/vlang

Thanks.

--- a/GNUmakefile
+++ b/GNUmakefile
@@ -7,9 +7,12 @@ VC     ?= ./vc
 VEXE   ?= ./v
 VCREPO ?= https://github.com/vlang/vc
 TCCREPO ?= https://github.com/vlang/tccbin
+LEGACYREPO ?= https://github.com/macports/macports-legacy-support
 
 VCFILE := v.c
 TMPTCC := $(VROOT)/thirdparty/tcc
+LEGACYLIBS := $(VROOT)/thirdparty/legacy
+TMPLEGACY := $(LEGACYLIBS)/source
 TCCOS := unknown
 TCCARCH := unknown
 GITCLEANPULL := git clean -xf && git pull --quiet
@@ -36,6 +39,9 @@ endif
 ifeq ($(_SYS),Darwin)
 MAC := 1
 TCCOS := macos
+ifeq ($(shell expr $(shell uname -r | cut -d. -f1) \<= 15), 1)
+LEGACY := 1
+endif
 endif
 
 ifeq ($(_SYS),FreeBSD)
@@ -79,13 +85,13 @@ endif
 endif
 endif
 
-.PHONY: all clean check fresh_vc fresh_tcc check_for_working_tcc
+.PHONY: all clean check fresh_vc fresh_tcc fresh_legacy check_for_working_tcc
 
 ifdef prod
 VFLAGS+=-prod
 endif
 
-all: latest_vc latest_tcc
+all: latest_vc latest_tcc latest_legacy
 ifdef WIN32
 	$(CC) $(CFLAGS) -std=c99 -municode -w -o v1.exe $(VC)/$(VCFILE) $(LDFLAGS)
 	v1.exe -no-parallel -o v2.exe $(VFLAGS) cmd/v
@@ -93,6 +99,12 @@ ifdef WIN32
 	del v1.exe
 	del v2.exe
 else
+ifdef LEGACY
+	$(MAKE) -C $(TMPLEGACY)
+	$(MAKE) -C $(TMPLEGACY) PREFIX=$(realpath $(LEGACYLIBS)) CFLAGS=$(CFLAGS) LDFLAGS=$(LDFLAGS) install
+	rm -rf $(TMPLEGACY)
+	$(eval override LDFLAGS+=-L$(realpath $(LEGACYLIBS))/lib -lMacportsLegacySupport)
+endif
 	$(CC) $(CFLAGS) -std=gnu99 -w -o v1.exe $(VC)/$(VCFILE) -lm -lpthread $(LDFLAGS)
 	./v1.exe -no-parallel -o v2.exe $(VFLAGS) cmd/v
 	./v2.exe -o $(VEXE) $(VFLAGS) cmd/v
@@ -104,6 +116,7 @@ endif
 
 clean:
 	rm -rf $(TMPTCC)
+	rm -rf $(LEGACYLIBS)
 	rm -rf $(VC)
 
 ifndef local
@@ -148,11 +161,32 @@ else
 	@$(MAKE) --quiet check_for_working_tcc 2> /dev/null
 endif
 
+ifndef local
+latest_legacy: $(TMPLEGACY)/.git/config
+ifdef LEGACY
+	cd $(TMPLEGACY) && $(GITCLEANPULL)
+endif
+else
+latest_legacy:
+ifdef LEGACY
+	@echo "Using local legacysupport"
+endif
+endif
+
+fresh_legacy:
+	rm -rf $(LEGACYLIBS)
+	$(GITFASTCLONE) $(LEGACYREPO) $(TMPLEGACY)
+
 $(TMPTCC)/.git/config:
 	$(MAKE) fresh_tcc
 
 $(VC)/.git/config:
 	$(MAKE) fresh_vc
+
+$(TMPLEGACY)/.git/config:
+ifdef LEGACY
+	$(MAKE) fresh_legacy
+endif
 
 asan:
 	$(MAKE) all CFLAGS='-fsanitize=address,undefined'

--- a/vlib/time/time_nix.c.v
+++ b/vlib/time/time_nix.c.v
@@ -3,6 +3,7 @@
 // that can be found in the LICENSE file.
 module time
 
+#flag darwin -I@VEXEROOT/thirdparty/legacy/include/LegacySupport
 #include <time.h>
 #include <errno.h>
 

